ring/ring.go

```diff
@@ -34,21 +34,58 @@ const (
 	GetBufferSize = 5
 )
 
+type Options struct {
+	ReplicationFactor int
+	BufDescs          []InstanceDesc
+	BufHosts          []string
+	BufZones          []string
+}
+
+type Option func(opts *Options)
+
+func WithBuffers(bufDescs []InstanceDesc, bufHosts, bufZones []string) Option {
+	return func(opts *Options) {
+		opts.BufDescs = bufDescs
+		opts.BufHosts = bufHosts
+		opts.BufZones = bufZones
+	}
+}
+
+func WithReplicationFactor(replication int) Option {
+	return func(opts *Options) {
+		opts.ReplicationFactor = replication
+	}
+}
+
+func collectOptions(opts ...Option) *Options {
+	final := &Options{}
+	for _, opt := range opts {
+		opt(final)
+	}
+	return final
+}
```

Comment:
we had some benchmarks about not allocating on `Get`. The plan is not to use this GetWithOptions on the hot path, right? (but also maybe just returning an `Options` might do away with the allocation)

Reply:
I'll see if I can improve this since it'd be nice to clean it up:

```
ring/ring.go:46:18: leaking param: bufDescs
ring/ring.go:46:43: leaking param: bufHosts
ring/ring.go:46:53: leaking param: bufZones
ring/ring.go:47:14: opts does not escape
ring/ring.go:47:9: func literal escapes to heap
ring/ring.go:55:14: opts does not escape
ring/ring.go:55:9: func literal escapes to heap
ring/ring.go:60:21: opts does not escape
ring/ring.go:61:11: &Options{} escapes to heap
```